### PR TITLE
Use new WinRT APIs for SpecialFolders (#22664)

### PR DESCRIPTION
**Pull request by JeremyKuhne**
Ports #22664. This addresses #20782.

* Use new WinRT APIs for SpecialFolders

* Update tests

* Update test package

Fix test verification

* Clean up project files

* Add back test skip condition

* Add comments and fallback Windows dir.

### CODE REVIEW

File: dependencies.props [hunk 1]
```diff
@@ -158,7 +158,7 @@
     </StaticDependency>
 
     <StaticDependency Include="Microsoft.TargetingPack.Private.WinRT">
-      <Version>1.0.3</Version>
+      <Version>1.0.5</Version>
```
Comment (joperezr) [inline dependencies.props]: I thought you were going to remove this. It's not used so we should just get rid of it.

Comment (JeremyKuhne) [inline dependencies.props]: I replied on the original issue that I would follow up separately. It took long enough to get this change to work down level that I didn't want to add anything that wasn't strictly necessary to the change.

File: src/System.Runtime.Extensions/src/System/Environment.WinRT.cs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+using Windows.Storage;
+using System.IO;
+
+namespace System
+{
+    public static partial class Environment
+    {
+        private static string GetFolderPathCore(SpecialFolder folder, SpecialFolderOption option)
+        {
+            // For testing we'll fall back if the needed APIs aren't present.
```
Comment (stephentoub) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: This is for testing, but we're shipping it?  Is it possible someone could end up using this in production, either on purpose or accidentally?

File: src/System.Runtime.Extensions/src/System/Environment.WinRT.cs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+            //
+            // We're not honoring the special folder options (noverify/create) for a few reasons. One, most of the
+            // folders always exist (e.g. it is moot). Two, most locations are inaccessible from an appcontainer
+            // currently - making it impossible to answer the question of existence or create if necessary. Thirdly,
+            // the Win32 API would create these folders with very specific ACLs, which even in the cases we can create
+            // are a significant compat risk (trying to replicate internal Windows behavior- it is documented that they
+            // set specific ACLs, but not which ones).
+            if (ApiInformation.IsTypePresent("Windows.Storage.UserDataPaths"))
```
Comment (stephentoub) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: I asked this on the original PR, but I'd still be interested in understanding what the cost of this is and what kind of overhead it's adding to GetSpecialFolder calls, whether we should consider caching it or not, etc.  It's of course possible it's inconsequential, but it'd be good to know for sure.

File: src/System.Runtime.Extensions/src/System/Environment.WinRT.cs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+{
+    public static partial class Environment
+    {
+        private static string GetFolderPathCore(SpecialFolder folder, SpecialFolderOption option)
+        {
+            // For testing we'll fall back if the needed APIs aren't present.
+            //
+            // We're not honoring the special folder options (noverify/create) for a few reasons. One, most of the
+            // folders always exist (e.g. it is moot). Two, most locations are inaccessible from an appcontainer
+            // currently - making it impossible to answer the question of existence or create if necessary. Thirdly,
```
Comment (stephentoub) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: So most of the paths we're handing out will result in exceptions when callers try to use them?

File: src/System.Runtime.Extensions/src/System/Environment.WinRT.cs (hunk 2 above)
Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: Hopefully the tests I written for RS3 will catch accidental usage.  When all of our development infrastructure is on RS3+ then the fallback could be removed.

File: src/System.Runtime.Extensions/src/System/Environment.WinRT.cs (hunk 3 above)
Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: I didn't measure anything significant when I looked at it.

File: src/System.Runtime.Extensions/src/System/Environment.WinRT.cs (hunk 4 above)
Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: Yes, for now. In the future this will likely change.

**Comment by JeremyKuhne** [comment]
> I'm still not clear on the usefulness of the change if most of the paths it gives out aren't actually usable, but assuming that's already the state of the world, I guess it doesn't make things worse.

Using SHGetKnownFolderPath was a temporary thing until Windows got their plans settled with known folders. None of the existing special folder APIs made it into the WACK. Final plans came in pretty late (DCR) due to long term planning implications.

**Pull request merged**